Avoid SidecarContainers code path for non-sidecar pods

gjkim42 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind regression

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

This fixes a regression in the SidecarContainers feature by minimizing the impact of the new code path. Use the old code path for pods without restartable init containers, and apply the new code path only to pods with restartable init containers.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

xref: #126440 

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixed a regression in 1.29+ default configurations, where regular init containers may fail to start due to a temporary container runtime failure.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

/cc @liggitt @SergeyKanzhelev 
/priority critical-urgent

gjkim42 committed:
Avoid SidecarContainers code path for non-sidecar pods

This fixes a regression in the SidecarContainers feature by minimizing
the impact of the new code path. Use the old code path for pods without
restartable init containers, and apply the new code path only to pods
with restartable init containers.

liggitt commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -832,7 +832,7 @@ func (m *kubeGenericRuntimeManager) killContainersWithSyncResult(ctx context.Con
 	var termOrdering *terminationOrdering
 	// we only care about container termination ordering if the sidecars feature is enabled
-	if utilfeature.DefaultFeatureGate.Enabled(features.SidecarContainers) {
+	if types.HasRestartableInitContainer(pod) {
```
slight preference for 

```go
if utilfeature.DefaultFeatureGate.Enabled(features.SidecarContainers) && types.HasRestartableInitContainer(pod) {
```

SergeyKanzhelev replied:
+1 for being more explicit here

liggitt commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -862,7 +862,7 @@ func (m *kubeGenericRuntimeManager) computePodActions(ctx context.Context, pod *
 		if len(containersToStart) == 0 {
 			hasInitialized := false
-			if !utilfeature.DefaultFeatureGate.Enabled(features.SidecarContainers) {
+			if !types.HasRestartableInitContainer(pod) {
```
compute once at the top of computePodActions:
```go
handleInitContainerSidecars := utilfeature.DefaultFeatureGate.Enabled(features.SidecarContainers) && types.HasRestartableInitContainer(pod)
```

then use that throughout `computePodActions`

liggitt commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -1285,7 +1285,7 @@ func (m *kubeGenericRuntimeManager) SyncPod(ctx context.Context, pod *v1.Pod, po
 	}
 
-	if !utilfeature.DefaultFeatureGate.Enabled(features.SidecarContainers) {
+	if !types.HasRestartableInitContainer(pod) {
```
slight preference for 

```go
if !(utilfeature.DefaultFeatureGate.Enabled(features.SidecarContainers) && types.HasRestartableInitContainer(pod)) {
```

gjkim42 replied:
I just realized that this was a bit bad decision.

If we didn't check types.HasRestartableInitContainer(pod) here, we would remove this code path safely while removing the feature gate.

https://github.com/kubernetes/kubernetes/pull/129731/commits/2cb6899b31a33a4ce36ecc8694af63a1bde60db1

@liggitt 

Do you have a good idea to remove this redundant code path (logically it is a subset of the else code path.)

liggitt commented on pkg/kubelet/types/pod_update.go:
```diff
@@ -202,3 +204,17 @@ func IsRestartableInitContainer(initContainer *v1.Container) bool {
+// HasRestartableInitContainer returns true if the pod has any restartable init
+// container and the feature gate is enabled.
+func HasRestartableInitContainer(pod *v1.Pod) bool {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.SidecarContainers) {
```
Making this include the feature gate check is a little confusing (IsRestartableInitContainer doesn't consider the feature enablement) and could lead to callers thinking a pod doesn't have sidecars when it does.

I think it's better to move the feature gate check to the call sites (there's only three).

gjkim42 replied:
done.

SergeyKanzhelev commented:
> Fixes https://github.com/kubernetes/kubernetes/issues/126440

Please just mention the bug without saying "Fixes", otherwise it will close the  bug and we still want to fix the new code path

pacoxu commented:
/cc @kubernetes/sig-node-leads 
as a critical regression

pacoxu commented:
/triage accepted

pacoxu commented:
As Sidecar feature was promoted to beta in v1.29, we would backport this to v1.29-v1.31 later.

BTW, how to add a test for this? I met a similar case in https://github.com/kubernetes/kubernetes/pull/122778 and https://github.com/kubernetes/kubernetes/pull/122778/#discussion_r1737681804 provide a way to test it locally using a customized containerd.

gjkim42 commented:
> BTW, how to add a test for this? I met a similar case in #122778 and [discussion_r1737681804](https://github.com/kubernetes/kubernetes/pull/122778/#discussion_r1737681804) provide a way to test it locally using a customized containerd, and in that case, we make a `startContainer` timeout.

Thanks, will test it manually first.

Not sure we want the customized container runtime for the CI though.

gjkim42 commented:
manually tested with customized containerd (that fails to start container at first after 1 min)

```
...
NAME        READY   STATUS     RESTARTS   AGE
fixed-k8s   0/1     Init:0/1   0          56s
k8s-1.30    0/1     Init:0/1   0          56s
NAME        READY   STATUS     RESTARTS   AGE
fixed-k8s   0/1     Init:0/1   0          58s
k8s-1.30    0/1     Init:0/1   0          58s
NAME        READY   STATUS     RESTARTS   AGE
fixed-k8s   0/1     Init:0/1   0          60s
k8s-1.30    0/1     Init:0/1   0          60s
NAME        READY   STATUS                   RESTARTS   AGE
fixed-k8s   0/1     Init:RunContainerError   0          63s
k8s-1.30    0/1     Init:RunContainerError   0          63s
NAME        READY   STATUS                   RESTARTS   AGE
fixed-k8s   0/1     Init:0/1                 1          65s
k8s-1.30    0/1     Init:RunContainerError   0          65s
NAME        READY   STATUS                   RESTARTS   AGE
fixed-k8s   0/1     Init:0/1                 1          67s
k8s-1.30    0/1     Init:RunContainerError   0          67s
NAME        READY   STATUS                   RESTARTS   AGE
fixed-k8s   0/1     Init:0/1                 1          69s
k8s-1.30    0/1     Init:RunContainerError   0          69s
NAME        READY   STATUS                   RESTARTS   AGE
fixed-k8s   0/1     Init:0/1                 1          71s
k8s-1.30    0/1     Init:RunContainerError   0          71s
NAME        READY   STATUS                   RESTARTS   AGE
fixed-k8s   0/1     Init:0/1                 1          73s
k8s-1.30    0/1     Init:RunContainerError   0          73s
NAME        READY   STATUS                   RESTARTS   AGE
fixed-k8s   0/1     PodInitializing          0          75s
k8s-1.30    0/1     Init:RunContainerError   0          75s
NAME        READY   STATUS                   RESTARTS   AGE
fixed-k8s   1/1     Running                  0          77s
k8s-1.30    0/1     Init:RunContainerError   0          77s
...
```
```
$ k get nodes
NAME                 STATUS   ROLES           AGE   VERSION
kind-control-plane   Ready    control-plane   28m   v1.32.0-alpha.0.498+fcf616daa7f18a-dirty
kind-worker          Ready    <none>          28m   v1.30.4
kind-worker2         Ready    <none>          28m   v1.32.0-alpha.0.498+fcf616daa7f18a-dirty
```

liggitt commented:
This change lgtm. Is there a test we can add that reproduces https://github.com/kubernetes/kubernetes/issues/126440 that shows this resolves the issue?

gjkim42 commented:
> This change lgtm. Is there a test we can add that reproduces #126440 that shows this resolves the issue?

added a unit test for that.
Since it is difficult to reproduce it in the e2e test, I tested it manually.
https://github.com/kubernetes/kubernetes/pull/127162#issuecomment-2333839460

liggitt commented:
/lgtm
/approve
/hold for @SergeyKanzhelev review

liggitt commented:
/assign @SergeyKanzhelev

SergeyKanzhelev reviewed (approved):
/unhold
/approve
/lgtm

liggitt commented:
thanks, can you cherry-pick this to 1.29, 1.30, and 1.31?
